src/Iesi.Collections/Generic/SynchronizedSet.cs

```diff
@@ -29,8 +29,8 @@ public sealed class SynchronizedSet<T> : ISet<T>
 		/// <param name="basisSet">The <c>Set</c> object that this object will wrap.</param>
 		public SynchronizedSet(ISet<T> basisSet)
 		{
-			_basisSet = basisSet;
-			_syncRoot = ((ICollection)basisSet).SyncRoot;
+			_basisSet = basisSet ?? throw new ArgumentNullException(nameof(basisSet));
+			_syncRoot = basisSet is ICollection c ? c.SyncRoot : new object();
```

Comment:
I don't actually know what to do if basisSet does not implement `ICollection`.

Reply:
I see four options:

 - Keep failing, but with a better error than an invalid cast.
 - Synchronize only that `SynchronizeSet` instance, (your change). That creates a discrepancy with sets implementing `ICollection`, because these ones will be synchronized across any `SynchronizeSet` instance build on them.
 - Synchronize on the `basisSet` instance as a sync root. Avoids above discrepancy. And [it appears](https://docs.microsoft.com/en-us/dotnet/api/system.collections.icollection.syncroot#remarks) that yielding the collection instance itself as `SyncRoot` can be an usual implementation.
 - Obsolete the whole type.

The third option looks better to me than the second one.

The first option causes the class to be quite useless for modern .Net applications. Better go for the fourth option then

So I think the choice has to be made between using `basisSet` as `_syncRoot` when `basisSet` is not an `ICollection` (and implement `ICollection` on `SynchronizeSet`, otherwise building a `SynchronizeSet` on a `SynchronizeSet` will not ensure synchronization across them), or obsolete the whole type and let it die someday.

Reply:
Ok, I've chosen to synchronize on `this` if `SyncRoot` is not available